### PR TITLE
MULE-16479: Expired Cookie passed to endpoint if flow starts from another Mule API.

### PR DESCRIPTION
**Pull request by emariotti3**

### CODE REVIEW

File: src/main/java/com/ning/http/client/RequestBuilderBase.java [hunk 1]
```diff
@@ -386,10 +386,14 @@ public T addOrReplaceCookie(Cookie cookie) {
 
             index++;
         }
-        if (replace)
+        if (cookie.hasExpired()) {
```
Comment (eze210) [inline src/main/java/com/ning/http/client/RequestBuilderBase.java]: Could this check be done at the beginning of the method?

File: src/main/java/com/ning/http/client/cookie/Cookie.java [hunk 2]
```diff
@@ -98,9 +100,13 @@ private static String validateValue(String name, String value) {
     private final boolean wrap;
     private final String domain;
     private final String path;
-    private final long maxAge;
+    private long maxAge = MAX_AGE_UNSPECIFIED;
```
Comment (eze210) [inline src/main/java/com/ning/http/client/cookie/Cookie.java]: It's always set in the constructor, so it can be made final again

File: src/main/java/com/ning/http/client/cookie/Cookie.java [hunk 3]
```diff
@@ -111,6 +117,7 @@ public Cookie(String name, String value, boolean wrap, String domain, String pat
         this.maxAge = maxAge;
         this.secure = secure;
         this.httpOnly = httpOnly;
+        this.whenCreated = System.currentTimeMillis();
```
Comment (eze210) [inline src/main/java/com/ning/http/client/cookie/Cookie.java]: Static import

File: src/test/java/com/ning/http/client/async/AsyncProvidersBasicTest.java [hunk 4]
```diff
@@ -464,6 +464,42 @@ public Response onCompleted(Response response) throws Exception {
...
+                        assertEquals(cookies.get(0).toString(), "cookie1=myCookie");
+                    } finally {
+                        l.countDown();
+                    }
+                    return response;
+                }
+            }).get();
+
+            if (!l.await(TIMEOUT, TimeUnit.SECONDS)) {
+                Assert.fail("Timeout out");
```
Comment (eze210) [inline src/test/java/com/ning/http/client/async/AsyncProvidersBasicTest.java]: Static import

File: src/test/java/com/ning/http/client/async/PerRequestRelative302Test.java [hunk 5]
```diff
@@ -105,6 +124,24 @@ public void redirected302Test() throws Throwable {
...
+        try (AsyncHttpClient client = getAsyncHttpClient(null)) {
+            // once
+            Response response = client.prepareGet(getTargetUrl()).setFollowRedirects(true).setHeader("X-redirect", getTargetUrl()).addQueryParam(TEST_COOKIES_KEY, TEST_COOKIES_VALUE).execute().get();
+
+            assertNotNull(response);
+            assertEquals(response.getStatusCode(), 200);
+
+            assertNotNull(cookies);
+            assertTrue(Arrays.asList(cookies).size() == 1);
+            Cookie cookie = Arrays.asList(cookies).get(0);
```
Comment (eze210) [inline src/test/java/com/ning/http/client/async/PerRequestRelative302Test.java]: Static import

File: src/main/java/com/ning/http/client/providers/grizzly/AsyncHttpClientFilter.java [hunk 6]
```diff
@@ -474,22 +475,28 @@ private void addAcceptHeaders(final HttpRequestPacket requestPacket) {
...
-            convertCookies(cookies, gCookies);
+            org.glassfish.grizzly.http.Cookie[] gCookies = new org.glassfish.grizzly.http.Cookie[finalCookiesList.size()];
+            convertCookies(finalCookiesList, gCookies);
             CookieSerializerUtils.serializeClientCookies(sb, gCookies);
             requestPacket.addHeader(Header.Cookie, sb.toString());
         }
     }
 
-    private void convertCookies(final Collection<Cookie> cookies, final org.glassfish.grizzly.http.Cookie[] gCookies) {
+    private void convertCookies(final List<org.glassfish.grizzly.http.Cookie> cookies, final org.glassfish.grizzly.http.Cookie[] gCookies) {
```
Comment (eze210) [inline src/main/java/com/ning/http/client/providers/grizzly/AsyncHttpClientFilter.java:493]: This method was used to convert from com.ning.http.client.cookie.Cookie to org.glassfish.grizzly.http.Cookie.
It's no longer needed since you are doing this conversion in the loop that filters the expired cookies

File: src/main/java/com/ning/http/client/RequestBuilderBase.java (hunk 1 above)
Comment (eze210) [reply src/main/java/com/ning/http/client/RequestBuilderBase.java]: Also, please double-check if this filter is ok in this method

**Review by eze210 — CHANGES_REQUESTED** [review]

File: src/main/java/com/ning/http/client/cookie/Cookie.java [hunk 7]
```diff
@@ -137,6 +144,20 @@ public String getPath() {
...
         return Long.MIN_VALUE;
     }
+
+    public boolean hasExpired() {
+        if (maxAge == 0) return true;
+
+        // if not specify max-age, this cookie should be
+        // discarded when user agent is to be closed, but
+        // it is not expired.
+        if (maxAge == MAX_AGE_UNSPECIFIED){
```
Comment (fsgonz) [inline src/main/java/com/ning/http/client/cookie/Cookie.java]: space after condition

File: src/main/java/com/ning/http/client/cookie/Cookie.java [hunk 8]
```diff
@@ -137,6 +144,20 @@ public String getPath() {
     public long getExpires() {
         return Long.MIN_VALUE;
     }
+
+    public boolean hasExpired() {
+        if (maxAge == 0) return true;
+
+        // if not specify max-age, this cookie should be
```
Comment (fsgonz) [inline src/main/java/com/ning/http/client/cookie/Cookie.java]: I would comment: "If max-age is not specified, this cookie should be discarded when user agent is to be closed but it is not expired"

File: src/test/java/com/ning/http/client/async/AsyncProvidersBasicTest.java [hunk 9]
```diff
@@ -441,7 +441,7 @@ public void asyncDoGetCookieTest() throws Throwable {
             h.add("Test4", "Test4");
             h.add("Test5", "Test5");
 
-            final Cookie coo = new Cookie("foo", "value", false, "/", "/", -1L, false, false);
+            final Cookie coo = new Cookie("foo", "value", false, "/", "/", Long.MIN_VALUE, false, false);
```
Comment (fsgonz) [inline src/test/java/com/ning/http/client/async/AsyncProvidersBasicTest.java]: static import if possible.

File: src/main/java/com/ning/http/client/cookie/Cookie.java [hunk 10]
```diff
@@ -137,6 +144,20 @@ public String getPath() {
...
+        // discarded when user agent is to be closed, but
+        // it is not expired.
+        if (maxAge == MAX_AGE_UNSPECIFIED){
+            return false;
+        }
+
+        long deltaSecond = (System.currentTimeMillis() - whenCreated) / 1000;
```
Comment (fsgonz) [inline src/main/java/com/ning/http/client/cookie/Cookie.java]: static import

File: src/main/java/com/ning/http/client/RequestBuilderBase.java [hunk 11]
```diff
@@ -373,23 +373,28 @@ public T addCookie(Cookie cookie) {
...
-        int index = 0;
+    public void cleanExpiredCookies() {
         lazyInitCookies();
+
         for (Cookie c : request.cookies) {
-            if (c.getName().equals(cookieKey)) {
-                replace = true;
-                break;
+            if (c.hasExpired()) {
+                request.cookies.remove(c);
```
Comment (eze210) [inline src/main/java/com/ning/http/client/RequestBuilderBase.java]: I think it could throw a ConcurrentModificationException because you are removing elements while iterating the container

File: src/main/java/com/ning/http/client/cookie/Cookie.java [hunk 12]
```diff
@@ -98,9 +102,13 @@ private static String validateValue(String name, String value) {
     private final boolean wrap;
     private final String domain;
     private final String path;
-    private final long maxAge;
+    private long maxAge;
```
Comment (eze210) [inline src/main/java/com/ning/http/client/cookie/Cookie.java]: Make it final

File: src/main/java/com/ning/http/client/RequestBuilderBase.java [hunk 13]
```diff
@@ -373,23 +373,28 @@ public T addCookie(Cookie cookie) {
         return derived.cast(this);
     }
 
-    public T addOrReplaceCookie(Cookie cookie) {
-        String cookieKey = cookie.getName();
-        boolean replace = false;
-        int index = 0;
+    public void cleanExpiredCookies() {
```
Comment (eze210) [inline src/main/java/com/ning/http/client/RequestBuilderBase.java:376]: Should this method be called from addCookie too?

**Review by eze210 — CHANGES_REQUESTED** [review]

File: src/main/java/com/ning/http/client/cookie/Cookie.java (hunk 7 above)
Comment (fsgonz) [reply src/main/java/com/ning/http/client/cookie/Cookie.java]: Is there consistency everywhere in grizzly that the maxAge is Long.MIN_VALUE? Previously the test set -1. Verify that.

File: src/main/java/com/ning/http/client/RequestBuilderBase.java [hunk 14]
```diff
@@ -373,23 +373,28 @@ public T addCookie(Cookie cookie) {
...
+        lazyInitCookies();
+
+        cleanExpiredCookies();
+        if (request.cookies.contains(cookie)) {
+            index = request.cookies.indexOf(cookie);
         }
-        if (replace)
-            request.cookies.set(index, cookie);
-        else
+        // if cookie not found, then add it
```
Comment (fsgonz) [inline src/main/java/com/ning/http/client/RequestBuilderBase.java]: shouldn't this be in an else as the -1 is changed only if request.cookies.contains(cookie) is false?

File: src/test/java/com/ning/http/client/async/AsyncProvidersBasicTest.java [hunk 15]
```diff
@@ -464,6 +464,42 @@ public Response onCompleted(Response response) throws Exception {
...
+
+            final Cookie cookie = new Cookie("cookie1", "myCookie", false, "/", "/", Long.MIN_VALUE, false, false);
+            final Cookie expiredCookie = new Cookie("cookie2", "myExpCookie", false, "/", "/", 0, false, false);
+
+            client.prepareGet(getTargetUrl()).setHeaders(h).addCookie(cookie).addCookie(expiredCookie).execute(new AsyncCompletionHandlerAdapter() {
+
+                @Override
+                public Response onCompleted(Response response) throws Exception {
+                    try {
+                        assertEquals(response.getStatusCode(), 200);
```
Comment (fsgonz) [inline src/test/java/com/ning/http/client/async/AsyncProvidersBasicTest.java]: use constants from a class for the standard response codes.

File: src/test/java/com/ning/http/client/async/AsyncProvidersBasicTest.java [hunk 16]
```diff
@@ -464,6 +464,42 @@ public Response onCompleted(Response response) throws Exception {
...
+                        assertEquals(cookies.size(), 1);
+                        assertEquals(cookies.get(0).toString(), "cookie1=myCookie");
+                    } finally {
+                        l.countDown();
+                    }
+                    return response;
+                }
+            }).get();
+
+            if (!l.await(TIMEOUT, TimeUnit.SECONDS)) {
```
Comment (fsgonz) [inline src/test/java/com/ning/http/client/async/AsyncProvidersBasicTest.java]: Will this fail always? Notice that the l.countdown is in the finally. Can't this be refactored using some kind of polling in grizzly?

**Review by fsgonz — CHANGES_REQUESTED** [review]

File: src/test/java/com/ning/http/client/async/AsyncProvidersBasicTest.java [hunk 17]
```diff
@@ -464,6 +464,42 @@ public Response onCompleted(Response response) throws Exception {
         }
     }
 
+    @Test(groups = { "standalone", "default_provider", "async" })
+    public void asyncDoGetCookieExpiredCookiesRemovedTest() throws Throwable {
+        try (AsyncHttpClient client = getAsyncHttpClient(null)) {
+            final CountDownLatch l = new CountDownLatch(1);
+            FluentCaseInsensitiveStringsMap h = new FluentCaseInsensitiveStringsMap();
```
Comment (fsgonz) [inline src/test/java/com/ning/http/client/async/AsyncProvidersBasicTest.java]: These headers don't seem to be related to the test.

File: src/test/java/com/ning/http/client/async/AsyncProvidersBasicTest.java [hunk 18]
```diff
@@ -464,6 +464,42 @@ public Response onCompleted(Response response) throws Exception {
...
+            final Cookie cookie = new Cookie("cookie1", "myCookie", false, "/", "/", Long.MIN_VALUE, false, false);
+            final Cookie expiredCookie = new Cookie("cookie2", "myExpCookie", false, "/", "/", 0, false, false);
+
+            client.prepareGet(getTargetUrl()).setHeaders(h).addCookie(cookie).addCookie(expiredCookie).execute(new AsyncCompletionHandlerAdapter() {
+
+                @Override
+                public Response onCompleted(Response response) throws Exception {
+                    try {
+                        assertEquals(response.getStatusCode(), 200);
+                        List<Cookie> cookies = response.getCookies();
```
Comment (fsgonz) [inline src/test/java/com/ning/http/client/async/AsyncProvidersBasicTest.java]: I would extract the asserts from here. Not using the finally and performing the asserts from outside the onComplete callback so that the test presents descriptive information in case of failure.

**Review by fsgonz — CHANGES_REQUESTED** [review]

File: src/test/java/com/ning/http/client/async/PerRequestRelative302Test.java [hunk 19]
```diff
@@ -105,6 +124,24 @@ public void redirected302Test() throws Throwable {
...
+        isSet.getAndSet(false);
+        try (AsyncHttpClient client = getAsyncHttpClient(null)) {
+            // once
+            Response response = client.prepareGet(getTargetUrl()).setFollowRedirects(true).setHeader("X-redirect", getTargetUrl()).addQueryParam(TEST_COOKIES_KEY, TEST_COOKIES_VALUE).execute().get();
+
+            assertNotNull(response);
+            assertEquals(response.getStatusCode(), 200);
+
+            assertNotNull(cookies);
+            assertTrue(asList(cookies).size() == 1);
```
Comment (fsgonz) [inline src/test/java/com/ning/http/client/async/PerRequestRelative302Test.java]: use assert equals

File: src/test/java/com/ning/http/client/async/PerRequestRelative302Test.java [hunk 20]
```diff
@@ -105,6 +124,24 @@ public void redirected302Test() throws Throwable {
...
 
+    @Test(groups = { "online", "default_provider" })
+    public void redirected302WithCookiesTest() throws Throwable {
+        isSet.getAndSet(false);
+        try (AsyncHttpClient client = getAsyncHttpClient(null)) {
+            // once
+            Response response = client.prepareGet(getTargetUrl()).setFollowRedirects(true).setHeader("X-redirect", getTargetUrl()).addQueryParam(TEST_COOKIES_KEY, TEST_COOKIES_VALUE).execute().get();
+
+            assertNotNull(response);
+            assertEquals(response.getStatusCode(), 200);
```
Comment (fsgonz) [inline src/test/java/com/ning/http/client/async/PerRequestRelative302Test.java]: use constant for status code like in the other cases.

**Review by fsgonz — CHANGES_REQUESTED** [review]

File: src/test/java/com/ning/http/client/async/PerRequestRelative302Test.java [hunk 21]
```diff
@@ -64,6 +81,8 @@ public void handle(String s, Request r, HttpServletRequest httpRequest, HttpServ
                     return;
                 }
             }
+
+            cookies = httpRequest.getCookies();
             httpResponse.setStatus(200);
```
Comment (fsgonz) [inline src/test/java/com/ning/http/client/async/PerRequestRelative302Test.java:87]: use constant

**Review by fsgonz — APPROVED** [review]

**Review by eze210 — APPROVED** [review]
